src/bsoncxx/json.cpp

```diff
@@ -0,0 +1,267 @@
+// Copyright 2014 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <bsoncxx/json.hpp>
+
+#include <iomanip>
+#include <iostream>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include <bsoncxx/document/view.hpp>
+#include <bsoncxx/stdx/string_view.hpp>
+#include <bsoncxx/types.hpp>
+#include <bsoncxx/types/value.hpp>
+
+#include <bsoncxx/b64_ntop.h>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+
+namespace {
+
+    class json_visitor {
+    public:
+        json_visitor(std::ostream& out, bool is_array, std::size_t padding)
+            : out(out), stack({is_array}), padding(padding) {}
+
+        void visit_key(stdx::string_view value) {
+            pad();
+
+            if (!stack.back()) {
+                out << "\"" << value.data() << "\""
+                    << " : ";
+            }
+        }
+
+        void visit_value(const types::b_eod&) {}
+
+        void visit_value(const types::b_double& value) { out << value.value; }
+
+        void visit_value(const types::b_utf8& value) { out << "\"" << value.value.data() << "\""; }
+
+        void visit_value(const types::b_document& value) {
+            out << "{" << std::endl;
+            stack.push_back(false);
+            visit_children(value.value);
+            pad();
+            out << "}";
+        }
+
+        void visit_value(const types::b_array& value) {
+            out << "[" << std::endl;
+            stack.push_back(true);
+            visit_children(value.value);
+            pad();
+            out << "]";
+        }
+
+        void visit_value(const types::b_binary& value) {
+
+            std::size_t b64_len;
+
+            b64_len = (value.size / 3 + 1) * 4 + 1;
```

Comment:
the size computation?

It's how large things can get after you run them through ntop.  You started with binary and you're going to a string encoding, so you get a growth rate associated with going from 8bits/byte to 6bits/byte.  I think I cribbed that particular arithmetic from somewhere reputable, but I can't quite remember where now.